torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -452,62 +462,25 @@ class TensorExprFuser {
   }
 #undef REQ
 
-  void fillTypeInfoMap(Block* block) {
-    for (Node* n : block->nodes()) {
-      if (n->kind() == prim::profile) {
-        if (auto tensor_ty = n->ty(attr::profiled_type)->cast<TensorType>()) {
-          GRAPH_DEBUG(
-              "TYPEINFO ADD: %", n->input()->debugName(), " --> ", *tensor_ty);
-          typeinfo_map_[n->input()] = tensor_ty;
-        }
-      }
-
-      for (Block* b : n->blocks()) {
-        fillTypeInfoMap(b);
-      }
-    }
-  }
-  void printTypeInfoMap() {
-    GRAPH_DEBUG("Typeinfo map:");
-    for (const auto& kv : typeinfo_map_) {
-      // GRAPH_DEBUG("%", kv.first->debugName(), " --> ", *kv.second);
-      GRAPH_DEBUG("%??? --> ", *kv.second);
-    }
-  }
-
-  void removeProfilingNodes(Block* b) {
-    for (auto it = b->nodes().begin(); it != b->nodes().end(); it++) {
-      if (it->kind() == prim::profile) {
-        GRAPH_DEBUG("TYPEINFO ERASE: %", it->output()->debugName());
-        typeinfo_map_.erase(it->output());
-        it->output()->replaceAllUsesWith(it->input());
-        it.destroyCurrent();
-      } else {
-        for (Block* ib : it->blocks()) {
-          removeProfilingNodes(ib);
-        }
-      }
-    }
-  }
-
   void guardFusionGroup(Node* fusion_group) {
     GRAPH_DEBUG("Inserting a typecheck guard for a node", *fusion_group);
     auto subgraph = SubgraphUtils::getSubgraph(fusion_group);
 
     // Fixup types of the subgraph inputs
     std::vector<Value*> inputs_to_check;
-    for (size_t idx = 0; idx < subgraph->inputs().size(); idx++) {
-      // TODO: If a type of the input was explicitly set in IR (in constrast to
-      // us expecting it from a hint given by a prim::profile node), we don't
-      // need to check that input in the typecheck.
-
+    for (Value* input : fusion_group->inputs()) {
       // We only check inputs of the fusion group and expect NNC to infer
       // intermediates and outputs shapes
-      if (typeinfo_map_.count(subgraph->inputs()[idx])) {
-        subgraph->inputs()[idx]->setType(
-            typeinfo_map_.at(subgraph->inputs()[idx]));
-        inputs_to_check.push_back(fusion_group->input(idx));
+      if (!input->type()->cast<TensorType>()) {
+        continue;
+      }
+
+      // fusion outputs are already guarded
+      if (input->node()->kind() == prim::Constant ||
```

Comment:
Thinking aloud: in future it might be worth adding an optimization to find redundant typechecks and remove them.